cEP-0022.md

```diff
@@ -0,0 +1,162 @@
+# Quickstart Green Mode
+
+| Metadata |                                               |
+| -------- | --------------------------------------------- |
+| cEP      | 22                                            |
+| Version  | 1.0                                           |
+| Title    | Quickstart Green Mode                         |
+| Authors  | Ishan Srivastava <mailto:ishan.srt@gmail.com> |
+| Status   | Implementation Due                            |
+| Type     | Feature                                       |
+
+# Abstract
+
+This document describes what is the **Green Mode** for `coala-quickstart` and how
+will this be implemented.
+
+This cEP describes new type of config files which will be generated by
+`coala-quickstart`, how will they be different from the ones currently being
+created, how is this helpful and the action-plan for adding this feature.
+
+# Introduction
+
+Currently `coala-quickstart` generates config files which point out too many
+errors in the code-base of the project. It also asks too many questions
+while generating the config file. The config files are not that specific to
+the project and the errors pointed out by coala using generated config files
+are generally undesirable for the project owners.
+
+Project owners need a configuration file which *agrees* with their codestyle, is
+as specific as possible i.e. can point out maximum errors with least amount
+of manual edits in the future, for future inconsistencies in commits.
+Many organizations don't follow specific code styles like `PEP8`, instead they
+follow their own standards. We need to adapt to those standards with least
+user interaction and produce config files which generate absolutely no errors
+in their codebase. This will lead to easier adoption of coala across communities
+as it is going to result in an easy mergeable Pull Request regarding
+*adding `.coafile` and CI to enforce it* in the other orgs for example
+<https://github.com/processing/processing/pull/4781>.
+
+# Implementation
+
+This project will be implemented in two phases:
+
+- The Brute Force way
+- Further optimization on the Brute Force
+
+1. The first step will consist of preparing a dict of all the values that can
```

Comment:
How are you going to prepare this dict?

Reply:
Some bear settings might have value in integral domain (like 1 to infinity), any strategies for that?

Reply:
Improving the metadata about settings belongs in the settings themselves.
c.f. https://github.com/coala/projects/issues/550

Adding missing defaults will help with the initial brute force attempt - with hundreds of settings that have no defaults, and many where a default would be very helpful, the combination is likely to mean that the brute force algorithm cant easily find a good initial set of settings to use as a basis to iteratively obtain better settings.

https://www.python.org/dev/peps/pep-0484/ must be understood.  The annotations need to become more precise types.  e.g. `typed_list(str)` in most places should be replaced with an `Enum` of some sort.

https://stackoverflow.com/questions/39398138/what-is-correct-way-of-type-hint-a-function-that-return-only-a-specific-set-of-v

These type annotations can then be loaded by quickstart and used to determine allowable values.  e.g. when it is an `Enum`, quickstart can get the list of acceptable values from the type.

For more complex rules, it would probably be a good idea to jump in the deep end and use type annotations by `contracts` , and then use its http://andreacensi.github.io/contracts/api_lowlevel.html to either eliminate potential values, or parse the given constraints to determine the entire range of allowable values.